text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,264 @@
+---
+rfc pr: [#290](https://github.com/aws/aws-cdk-rfcs/pull/290)
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/287
+---
+
+# CDK CLI deprecation warnings
+
+If an element in the CDK Construct Library (class, interface, property, function) is deprecated,
+using it inside a CDK application should produce a warning in the CDK CLI for all commands that perform synthesis
+(`cdk synth`, `cdk diff`, `cdk deploy`, etc.).
+
+This is done in order to aid the migration from `V1` of the CDK to `V2`
+(where all deprecated elements from `V1` will be removed).
+
+## Working Backwards
+
+### CHANGELOG
+
+feat(cli): warn about usages of deprecated Construct Library elements
+
+### README
+
+When using any element (class, interface, property, function) that is deprecated in the Construct Library,
+you will receive a warning when performing any CLI operation that performs synthesis
+(`cdk synth`, `cdk diff`, `cdk deploy`, etc.) that looks similar to:
+
+```
+[WARNING] The API @aws-cdk/core.CfnInclude is deprecated:
+  please use the CfnInclude class from the cloudformation-include module instead.
+  This API will be removed in the next major release
+```
+
+There is also an environment variable `CDK_DEPRECATED` that you can use to change the behavior of this feature:
```

Comment:
Since this is implemented at the JSII level it doesn't make sense to call this `CDK_`, either `JSII_DEPRECATED` or just `DEPRECATED`

```suggestion
There is also an environment variable `DEPRECATED` that you can use to change the behavior of this feature:
```

Reply:
> Since this is implemented at the JSII level it doesn't make sense to call this `CDK_`,

Yes. That's why, a little below, it says:

> #### 1. Changes to JSII
>
> (...)
> We will also add an option to the JSII CLI that allows passing the name of the environment variable used for silencing the warnings, and turning the warnings into errors
> (the `CDK_DEPRECATED` environment variable mentioned above -
> it can't be just `CDK_DEPRECATED` by default, as JSII does not have any knowledge of CDK!).

I feel like that's a better solution than `JSII_DEPRECATED`, which is too low-level (I don't think many customers are aware of JSII, and TS/JS customers don't use it at all), and `DEPRECATED`, which is too generic.

Reply:
I feel like adding a CLI switch that says what the name of this environment variable is unneeded complexity. Let's keep things simple just call this `DEPRECATED`. It's not a one-way door and if we end up deciding that we want to add support for "controlling the name of the deprecation environment variable in jsii", we can always add that later (but I doubt that will ever happen).

Reply:
Ok. Changed.